sdv/single_table/base.py

```diff
@@ -42,3 +46,175 @@ def get_parameters(self):
     def get_metadata(self):
         """Return the ``SingleTableMetadata`` for this synthesizer."""
         return self.metadata
+
+    def _validate_metadata_matches_data(self, columns):
+        errors = []
+        metadata_columns = self.metadata._columns if self.metadata._columns else []
+        invalid_data_columns = set(columns).difference(metadata_columns)
+        if invalid_data_columns:
+            errors.append(
+                f'The columns {sorted(invalid_data_columns)} are not present in the metadata.')
+
+        invalid_metadata_columns = set(metadata_columns).difference(columns)
+        if invalid_metadata_columns:
+            errors.append(
+                f'The metadata columns {sorted(invalid_metadata_columns)} '
+                'are not present in the data.'
+            )
+
+        if errors:
+            raise InvalidDataError(errors)
+
+    def _get_primary_plus_alternate_keys(self):
+        keys = set()
+        if self.metadata._primary_key:
+            keys = {self.metadata._primary_key}
+
+        if self.metadata._alternate_keys:
+            keys.update(set(self.metadata._alternate_keys))
+
+        return keys
+
+    def _get_sequence_keys(self):
+        if isinstance(self.metadata._sequence_key, tuple):
+            return set(self.metadata._sequence_key)
+
+        if isinstance(self.metadata._sequence_key, str):
+            return {self.metadata._sequence_key}
+
+        return set()
+
+    def _validate_keys_dont_have_missing_values(self, data):
+        errors = []
+        keys = self._get_primary_plus_alternate_keys()
+        keys.update(self._get_sequence_keys())
+        for key in sorted(keys):
+            if pd.isna(data[key]).any():
+                errors.append(f"Key column '{key}' contains missing values.")
+
+        return errors
+
+    @staticmethod
+    def _update_invalid_values(invalid_values):
+        invalid_values = sorted(invalid_values, key=lambda x: str(x))
+        if len(invalid_values) > 3:
+            return invalid_values[:3] + [f'+ {len(invalid_values) - 3} more']
+
+        return invalid_values
+
+    def _validate_key_values_are_unique(self, data):
+        errors = []
+        keys = self._get_primary_plus_alternate_keys()
+        for key in sorted(keys):
+            repeated_values = set(data[key][data[key].duplicated()])
+            if repeated_values:
+                repeated_values = self._update_invalid_values(repeated_values)
+                errors.append(f"Key column '{key}' contains repeating values: {repeated_values}")
+
+        return errors
+
+    def _validate_context_columns(self, data):
+        errors = []
+        context_column_names = self._data_processor._model_kwargs.get('context_columns')
```

Comment:
This is currently incorrect. I was unsure where we want to put it (inside the PAR model I presume?), so I just implemented/tested it here.